### DISCUSSION TITLE
Number of Params Actually Passed In

### DISCUSSION THREAD

**Discussion by irinaleo** [remkop/picocli#2078] [Q&A] [answered]
Hi,

With an interactive shell where I have @Command methods, is there something built-in to get the number of parameters actually passed in? Manually, I can check each param for non-null and add to a counter, but this seems like something picocli could take care of.

For example,

     @Command(name = "subcommand", description = "Subcommand")
     public void subcommand(
            @Parameters(arity = "1", description = "Parameter 1") String parameter1,
            @Parameters(arity = "0..1", description = "Optional parameter 2") String parameter2,
            @Parameters(arity = "0..1", description = "Optional parameter 3") String parameter3,
            @Parameters(arity = "0..1", description = "Optional parameter 4") String parameter4) {
    
             if (parameterCount > 3) {
                      // do something
             }
      }

I need it to not affect how the method is called i.e. adding another parameter.
I tried adding 'CommandLine.ParseResult parseResult' as a param (suggestion from ChatGPT) and parseResult.expandedArgs().size() seems to hold this info, but I can't run or test with it since it tells me I'm missing a param or  No TypeConverter is registered.

**Comment by remkop** [comment, marked as answer]
You may be able to get a `ParseResult` instance like this:

```java
class MyClass {

  @Spec CommandSpec spec;

 @Command(name = "subcommand", description = "Subcommand")
 public void subcommand(
        @Parameters(arity = "1", description = "Parameter 1") String parameter1,
        @Parameters(arity = "0..1", description = "Optional parameter 2") String parameter2,
        @Parameters(arity = "0..1", description = "Optional parameter 3") String parameter3,
        @Parameters(arity = "0..1", description = "Optional parameter 4") String parameter4) {

        ParseResult pr = spec.subcommands().get("subcommand").getParseResult();
        int parameterCount  = pr.expandedArgs().size();

         if (parameterCount > 3) {
                  // do something
         }
  }
```

**Reply by irinaleo** [reply]
I had to make spec static, otherwise this works. Not very intuitive, but it works!

**Reply by remkop** [reply]
Really, why does the spec field need to be static? In your example, the `@Command`-annotated method was an instance method, but yes, if in your application that method is static, then the spec field needs to be static to be used from the static method.

I see your point about this solution not being immediately obvious but I am not sure how to improve that.
The `@Spec` annotation is mentioned in the user manual ([this example](https://picocli.info/#_was_a_value_defaulted) is close to your use case), and I was hoping that that would give most users enough opportunity to find out about this.
But if you have suggestions to improve things please let me know.
